### PR TITLE
hit: create a hit syntax formatter

### PR DESCRIPTION
**Pull request by rwcarlsen**
This creates a formatting capability (via hit::Formatter) that allows auto-formatting hit input texts into a specified canonical format.  The formatter has a default configuration that uses two-space indent, 100-char line breaking and removes "./" and "../" from section headers and footers respectively.  Each of these defaults can be changed.  The formatter can also be configured to sort parameters/sections.

This PR just creates the formatter and corresponding python bindings to it.  A follow-on PR will include moose-integration and moose docs for using it.

ref #10845

### CODE REVIEW

File: framework/contrib/hit/parse.cc
Comment (permcody) [inline framework/contrib/hit/parse.cc]: Hmm... No errors on 4.8.4

File: framework/contrib/hit/parse.cc
Comment (permcody) [inline framework/contrib/hit/parse.cc]: This method definitely needs some doco. It's pretty difficult to understand. Looks like a quadratic algorithm? Hopefully nobody puts 10000 nodes in a tree!

File: framework/contrib/hit/parse.h
Comment (permcody) [inline framework/contrib/hit/parse.h]: Pedantic and minor: Maybe add some whitespace between your comment blocks and the subsequent method names?

**Review by permcody — DISMISSED** [review]
This all sits outside of MOOSE and I don't see any major issues. I'm approving. I'd like to see documentation on your sort method though...

File: framework/contrib/hit/parse.cc
Comment (rwcarlsen) [reply framework/contrib/hit/parse.cc]: Yes - it is quadratic, partly by necessity because later sorting rules can override earlier sort priorities - also because I am lazy and it is fast enough.  I ran it with sorting on the `moose_test --dump`output (multi-meg file) and it ran in ~10s of milliseconds.

**Review by friedmud — CHANGES_REQUESTED** [review]
Holding off on merging this until we see what the full MOOSE configuration file looks like.

**Comment by rwcarlsen** [comment]
I pushed up a change that tries out an example-based sort-order specification for the format style file.   A sort order spec would look something like this:

```

[format]
    indent_string = "  "
    line_length = 100
    canonical_section_markers = true

    [sorting]
        [.*]
          type = dummy
          [.*]
              type = dummy
          []
        []

        [Mesh]
          type = GeneratedMesh
          dim = 2
          ** = dummy
          nx = 10
          ny = 10
          nz = 10
        []

        [Variables]
        []

        [Kernels]
        []

        [BCs]
        []

        [**]
        []

        [Executioner]
        []

        [Outputs]
          exodus = true
        []
    []
[]
```

File: framework/contrib/hit/parse.cc
Comment (rwcarlsen) [reply framework/contrib/hit/parse.cc]: Actually - it isn't really quadratic.  It is `n*m` where n is the number of nodes and m is the number of order patterns (which should never be bigger than ~10s).

**Comment by friedmud** [comment]
I like the "example-based sort order"!  Well done!  You captured the best of your original design - and made it user-friendly.

This looks good.

**Comment by permcody** [comment]
Ping @friedmud - Not sure if your comment is an approve or if you had additional comments.

File: framework/contrib/hit/parse.cc
Comment (friedmud) [inline framework/contrib/hit/parse.cc]: What is this business?  Use `std::unique_ptr`

**Review by friedmud — DISMISSED** [review]
Ok - just the one comment.

Also: It pains me that you're doing your own command-line parsing.  I really wish we would use a utility for that (both here and in MOOSE!)... but I don't feel like ripping all that out at this moment.  Let's just get this in.

**Comment by rwcarlsen** [comment]
The standalone app (and its flag handling) are not how we will usually be formatting files in MOOSE - Moose will use the formatter API directly.

File: framework/contrib/hit/parse.cc
Comment (rwcarlsen) [reply framework/contrib/hit/parse.cc]: You don't need to free it.  The tree manages its own memory (manually).  But here the `delete` operation is actually removing nodes from the tree.

Comment (friedmud) [reply framework/contrib/hit/parse.cc]: We crossed comments there.

In modern times there is simply no reason for this pattern (to call a function and then free stuff).  If it would require a rewrite of HIT then that's what it requires (I doubt that though... I'm sure you can retrofit your current design with some smart pointers... even if they're `std::shared_ptr`).

If that's outside the scope of this PR then I can accept that - but I'd like to see an Issue created by you for removal of this pattern from HIT first.

Comment (permcody) [reply framework/contrib/hit/parse.cc]: OK - so we have a plan forward here:

I talked to Robert and he says that the internal tree structure manages it's own memory completely. Deleting a tree will delete all memory in the tree. What we'd like to do, is get to the point where you don't have to even delete the tree... This is important because with tree manipulation, any node in the tree has the potential to become a root which requires a delete:

1) Handing back memory that needs to be deleted is a no-go in 2018. John, David and I (as well as you) are all in agreement about this. A good compromise to rewriting all of his is to have any manipulation that "creates" or orphans a subtree (e.g. clone) will return back a smart pointer to that subtree. The internal nodes which are managed by their parent can remain dump pointers (I'll leave it up to Robert to decide which flavor). 

2) Any inconsistencies or other issues that arise from using a smart pointer to the root will be addressed. The end goal is external users should not have to alloc or delete memory. Instead they should create or remove nodes.

3) Move away from using the memory keyword "delete" as a tree manipulation function. Instead, let's make the Node API a little smarter and use a method instead.

If I've stated any of these bullets incorrectly, please correct them. Let's make sure we get the right compromise design moving forward.

At this time, rewriting the internal structure of the HIT tree to use all smart pointers is probably not a very good use of Robert's time. I do have other things I'd like him to focus on. I still think we should revisit this, perhaps in the future when we plan to move HIT outside of MOOSE or up into libMesh. that'll be a good time for an internal representation rewrite as we'll have probably learned quite a bit.

I think we may have an issue open for the internal representation rewrite. @rwcarlsen will you find that ticket and make any relevant design comments on that ticket? Let's not hold up this PR for the time being.

Comment (friedmud) [reply framework/contrib/hit/parse.cc]: 👍

**Review by permcody — APPROVED** [review]

**Pull request merged**
